[Tan Yong Feng Deon] iP

deeyonn opened the pull request:
# JennyBot Application
> "insert cliché programming related quote here" - Deon ([Browse](https://betterprogramming.pub/101-funny-programmer-quotes-76c7f335b92d))

JennyBot befriends you and helps you remember your tasks when no one else does.

## Features of JennyBot :hot_face::sweat_drops::sweat_drops:
- Remembers **three** different types of tasks:
  1. Todo task (you know, just a task)
  2. Deadline task (due by a certain date)
  3. Event task (occurring at a certain date)
- **Backup** your task list locally (user/.jenny/storage/tasks)
- Mark and unmark your tasks as **completed** or **uncompleted**
- **Finds** which tasks you are looking for

### List supported commands for JennyBot :wink:

1. Add a _Todo_ task `todo <description>`
2. Add a _Deadline_ task `deadline <description> /by <yyyy-MM-dd>`
3. Add a _Event_ task `event <description> /at <yyyy-MM-dd>`
4. List all available tasks `list`
5. Mark a task as complete `mark <index>`
6. Unmark a task as complete `unmark <index>`
7. Delete a task from the list `delete <index>`
8. Find a task by description `find <description>`
9. Exit the application `bye`

### Upcoming features from developers of JennyBot:

- [x] Using checkstyle to detect coding violations
```groovy
checkstyle {
    toolVersion = '10.2'
}
```
- [ ] Fresh slick Graphical User Interface

```java
import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.Label;
import javafx.stage.Stage;

public class JennyBot extends Application {

  // ...

  @Override
  public void start(Stage stage) {
    Label helloWorld = new Label("Hello World!"); // Creating a new Label control
    Scene scene = new Scene(helloWorld); // Setting the scene to be our Label

    stage.setScene(scene); // Setting the stage to show our screen
    stage.show(); // Render the stage.
  }
}
```

- [ ] Usage of variable arguments in JennyBot

```java
public String formatWithVarArgs(String... values) {
    // ...
}

formatWithVarArgs();
formatWithVarArgs("a", "b", "c", "d");
```

Commit:
Add Gradle support

deeyonn committed:
Add Level-1 implementation

deeyonn committed:
Added Helpers
Moved print_line() to Helpers
Moved print() to Helpers
Modified Duke to use Helpers
Added greet()

deeyonn committed:
Fixed typos

deeyonn committed:
Added implementation for Level-2
Modified Duke
Commented echo()
Added add()

deeyonn committed:
Refactored Helper to DukePrinter

deeyonn committed:
Changed Jenny chatbot back to Duke
Changed functions in Duke to static

deeyonn committed:
Implemented Level-3
Modified Duke logic
Moved all printing to DukePrinter
Created TaskItem for A-Classes

deeyonn committed:
Implemented Level-4
Modified Duke logic
Modified DukePrinter
Created DeadlineTask for A-Inheritance
Created TodoTask for A-Inheritance
Created EventTask for A-Inheritance

deeyonn committed:
Implemented A-TextUiTesting

deeyonn committed:
Moved DeadlineTask to TaskItems package
Moved EventTask to TaskItems package
Moved TaskItem to TaskItems package
Moved TodoTask to TaskItems package

Created DeadlineTaskException in TaskItems package
Created EventTaskException in TaskItems package
Created TodoTaskException in TaskItems package

Modified Duke logic to implement Level-5
Modified DukePrinter logic implement Level-5

deeyonn committed:
Modified Duke to implement Level-6
Modified DukePrinter to implement Level-6

deeyonn committed:
Refactor Duke.java to Jenny.java

There is no customisation of the application.

We need to customise the name of the application to have its own flavour.

Let's refactor the name of the application from Duke to Jenny.

deeyonn committed:
Refactor DukePrinter.java to JennyPrinter.java

There is no customisation of the application.

We need to customise the name of the application to have its own flavour.

Let's refactor the name of the application from Duke to Jenny.

deeyonn committed:
Refactor Duke to Jenny

There is no customisation of the application.

We need to customise the name of the application to have its own flavour.

Let's refactor the name of the application from Duke to Jenny.

deeyonn committed:
Add JennyStorage.java

There is no implementation for file I/O operations in the Jenny application.

We need to add this feature to save and load a list of tasks.

Let's add a class JennyStorage to handle file I/O operations of the Jenny application.

Implementation is currently work in progress.

deeyonn committed:
Modify DeadlineJennyTask.java, EventJennyTask.java, JennyTask.java, TodoJennyTask.java

There is no implementation for file I/O operations in the Jenny application.

We need to add this feature to save and load a list of tasks.

Let's add a function save() to format data for use in file I/O operations.

deeyonn committed:
Refactor various files related to TaskItem to JennyTask

There is no consistency in the naming of several class files in the application.

We need to maintain a certain level of consistency for ease of future development.

Let's refactor these files so that they are more consistent with what they represent.

deeyonn committed:
Refactor various files related to TaskItem to JennyTask

There is no consistency in the naming of several class files in the application.

We need to maintain a certain level of consistency for ease of future development.

Let's refactor these files so that they are more consistent with what they represent.

deeyonn committed:
Modify JennyBot.java and JennyStorage.java

There is no complete file I/O operation implementation for the JennyBot application.

We need to improve on the implementation for file I/O operations.

Let's add functions save() and load() to manage file I/O operations in the JennyBot application.

deeyonn committed:
Modify JennyBot.java and JennyStorage.java

There is no complete file I/O operation implementation for the JennyBot application.

We need to improve on the implementation for file I/O operations.

Let's add functions save() and load() to manage file I/O operations in the JennyBot application.

deeyonn committed:
Fix line indentations

deeyonn committed:
Refactor and repackage various files

* JennyTask refactored to AbstractTask, moved to jenny.tasks.
* DeadlineJennyTask refactored to Deadline, moved to jenny.tasks.
* EventJennyTask refactored to Event, moved to jenny.tasks.
* JennyPrinter refactored to Printer, moved to jenny.util.
* JennyStorage refactored to Storage, moved to jenny.storage.
* Removed DeadlineJennyTaskException.
* Removed EventJennyTaskException.
* Removed TodoJennyTaskException.
* Added TaskException to replace the 3 above, moved to jenny.exceptions.
* Moved JennyBot to jenny.

deeyonn committed:
Add Level-7 implementation

deeyonn committed:
Add DateTimeParser

deeyonn committed:
Add DateTimeParser

deeyonn committed:
Modify DateTimeParser

deeyonn committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

deeyonn committed:
Merge branch-Level-7 and branch Level-8

deeyonn committed:
Merge branch-Level-7 and branch Level-8

deeyonn committed:
Rename DateTimeParser to Validator

deeyonn committed:
Add JennyException

deeyonn committed:
Modify Validator

* Can now parse dates.
* Throws JennyException on parse failure.

deeyonn committed:
Add A-MoreOOP implementation

* Add AbstractCommand
* Add ByeCommand
* Add DeadlineCommand
* Add DeleteCommand
* Add EventCommand
* Add InvalidCommand
* Add ListCommand
* Add LoadCommand
* Add MarkCommand
* Add SaveCommand
* Add TodoCommand
* Add UnmarkCommand
* Modify AbstractStorage
* Modify AbstractTask
* Modify DeadlineTask
* Modify EventTask
* Modify JennyBot
* Modify JennyException
* Modify Printer
* Modify TaskStorage
* Modify TodoTask
* Modify Validator
* Add Parser
* Add UserInterface
* Moved various files to their respective packages

deeyonn committed:
Remove unused import packages

deeyonn committed:
Add implementation for A-MoreOOP and A-Packages

deeyonn committed:
Add JavaDoc for JennyBot

deeyonn committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

deeyonn committed:
Add files for A-Gradle

deeyonn committed:
Add files for checkstyle

deeyonn committed:
Add files for A-JUnit

deeyonn committed:
Modify build.gradle

* Removed duplicate code

deeyonn committed:
Add implementation for Level-9

* Add FindCommand
* Add FindCommandTest
* Modify Parser to use FindCommand
* Modify Task to have getDescription
* Modify TaskList to have filter

deeyonn committed:
Fix Java checkstyle violations

deeyonn committed:
Add JavaDoc for ByeCommand

deeyonn committed:
Merge branch 'branch-A-JavaDoc'

deeyonn committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/jenny/commands/ByeCommand.java

deeyonn committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/jenny/tasks/TaskList.java

deeyonn committed:
Fix checkstyle violations

deeyonn committed:
Modify README.md to use GFMD

jolynloh commented on src/main/java/jenny/JennyBot.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Run the JennyBot Application
```
Should it be 'Runs ...' instead of 'Run...'? ([Javadoc comments reference](https://se-education.org/guides/conventions/java/intermediate.html#comments))

deeyonn committed:
Modify Expected

jolynloh commented on src/main/java/jenny/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,64 @@
+package jenny.commands;
+
+import java.util.ArrayList;
+
+import jenny.exceptions.JennyException;
+import jenny.storage.Storage;
+import jenny.tasks.Task;
+import jenny.tasks.TaskList;
+import jenny.util.Ui;
```
I like how the import statements have a consistent ordering, making it easy to browse

jolynloh commented on src/main/java/jenny/commands/FindCommand.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    @Override
+    public void run(TaskList tasks, Ui ui, Storage<ArrayList<Task>> storage) throws JennyException {
+        ArrayList<Task> filtered = tasks.filter(description -> description.contains(arguments));
```
Perhaps using a plural form such as filteredTasks will better reflect that this is a collection that can contain more than one item?

jolynloh commented on src/main/java/jenny/tasks/TaskList.java:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @return the number of elements in this list
+     */
+    public int size() {
```
Consider using a verb such as `getSize()` for the method name instead? ([reference](https://se-education.org/guides/conventions/java/intermediate.html#naming))

jolynloh commented on src/main/java/jenny/util/Ui.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    /**
+     * Read and return a line from the input stream if present,
```
Should it be 'Reads and returns' instead?

jolynloh reviewed (commented):
LGTM, other than some nits. 
Coding standard is closely adhered to in most places 👍

deeyonn committed:
Modify build.gradle

* Set-up dependencies for JavaFX

deeyonn committed:
Add basic JavaFX feature

deeyonn committed:
Add support for echo in GUI

deeyonn committed:
Fix list command printing

list command is printing all tasks in one line.

list command should print each task at its own line.

Let's use regex and split the long string by ", " so that Printer.java can handle the rest.

Printing tasks in multiple line is preferable over one line for presentation and readability purposes.

deeyonn committed:
Add JavaDoc in Ui.java

One of the overloaded print functions in Ui.java does not have a JavaDoc.

All public methods should have their own JavaDoc describing what they do.

Let's write a JavaDoc for this method.

JavaDoc is essential for code maintenance and communicating purpose to future developers.

deeyonn committed:
Redirect I/O to use the new GUI

The application uses I/O from System.in and System.out and the new GUI only echos user input from the text box.

The application needs to make use of the I/O from the new GUI.

Let's:
* Create a new PrintStream in JennyBot::getResponse and set Ui.java to use it.
* Run the commands using user input from the text box in JennyBot::getResponse.
* Return the response from the new PrintStream as a String.

Redirection of I/O is much easier than overhauling the entire I/O process.

deeyonn committed:
Move MainWindow.java and DialogBox.java into package views

* Refactor JennyBot.java for the files moved.
* Refactor Main.java for the files moved.
* Refactor MainWindow.java for the files moved.
* Refactor MainWindow.fxml for the files moved.
* Modify JavaDoc for JennyBot::getResponse
* Modify access for JennyBot::getResponse

deeyonn committed:
Use assertions to verify assumptions

We do not use assertions to verify assumptions in our codebase.

We must not let certain assumptions for our code run free without
verification.

Let's use assertions to verify certain assumptions we have about our
code we write.

Assertions are suitable for verifying assumptions about Internal
Invariants, Control-Flow Invariants, Pre-conditions, Post-conditions,
and Class Invariants.

deeyonn committed:
Modify codebase to adhere to A-CodeQuality

Some parts of our codebase does not adhere to code quality guidelines.

Good code quality is essential to ease of readability and future
code maintenance.

Let's rewrite our codebase to fit the following guidelines.
* Be wary when a method is beyond 30 LOC (lines of code).
* If you need more than 3 levels of indentation, you're screwed.
* Avoid complicated expressions.
* Avoid magic numbers.
* Other guidelines at codeQuality.html#code-quality

Production code needs to be of high quality. Given how the world is
becoming increasingly dependent on software, poor quality code is
something no one can afford to tolerate.

deeyonn committed:
Modify JennyBot.java to use Streams for A-Streams.

deeyonn committed:
Merge pull request #1 from deeyonn/branch-A-Assertions

Add assertions to the codebase

deeyonn committed:
Merge pull request #2 from deeyonn/branch-A-CodeQuality

Modify codebase to adhere to code quality requirements

deeyonn committed:
Merge pull request #3 from deeyonn/branch-A-Streams

Modify JennyBot.java to use Streams for A-Streams.

deeyonn committed:
Remove temporary comment

deeyonn committed:
Modify GUI

* Modify pictures used for user and Jenny.
* Modify background colour for text.
* Modify text font and size.

deeyonn committed:
Remove ERROR_NOTHING_TO_SAVE message

deeyonn committed:
Add Greeting message to GUI

deeyonn committed:
Remove unused code after migrating to GUI

deeyonn committed:
Fix JavaDoc headers

deeyonn committed:
Fix operator with offset

deeyonn committed:
Remove unnecessary checks

deeyonn committed:
Modify application personality

deeyonn committed:
Add BCD-Extension

* Add feature to add notes

deeyonn committed:
Merge pull request #5 from deeyonn/branch-D-Notes

Add BCD-Extension

deeyonn committed:
Fix NoteCommand creating a wrong task.

deeyonn committed:
Add load function for NoteTask

deeyonn committed:
Fix missing break statement

deeyonn committed:
Add User Guide

deeyonn committed:
Modify application start class
